### PR TITLE
Add AES Key Wrap specification

### PR DESCRIPTION
**Pull request by bboston7**
This PR adds a specification for AES Key Wrap.  It shouldn't be merged until after the `aes` branch is merged.  I also wasn't sure whether `sha-imparative` or `master` was the right branch to target.  The only change I've made is to add the `AESKeyWrap.cry` file.  All other changes are from the `aes` branch.

### CODE REVIEW

File: Primitive/Symmetric/Cipher/Block/AESKeyWrap.cry [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+// Unwrap `ciphertext` using `key` as the key encryption key.  `iv` is the
+// expected initial value.  This function implements the algorithm from Section
+// 2.2.2 using the index-based definition.  It also checks the key data
+// integrity using `iv` (Section 2.2.3) and raises an error if the integrity
+// check fails.
+aesUnwrapKey : {n} (fin n, n >= 24, n % 8 == 0, width (n / 8) <= 64) =>
+               [AESKeySize] -> [8][8] -> [n][8] -> [n-8][8]
+aesUnwrapKey key iv ciphertext = if ((join iv) == A')
+                                 then split (join R')
+                                 else error "IV mismatch in decrypted ciphertext"
```
Comment (bboston7) [inline Primitive/Symmetric/Cipher/Block/AESKeyWrap.cry:46]: @andreistefanescu Do you think raising an error is appropriate for an integrity check failure here?  I also considered having this function return a tuple of the decrypted result and a boolean indicating whether the integrity check passed, or just skipping the integrity test altogether.

**Comment by weaversa** [comment]
In case you’re interested, we developed a pretty nice spec of all of NIST 800-38f for the Cryptol course, available here: https://github.com/weaversa/cryptol-course/tree/master/labs/KeyWrapping/spec/NIST_800_38F

File: Primitive/Symmetric/Cipher/Block/AESKeyWrap.cry (hunk 1 above)
Comment (weaversa) [reply Primitive/Symmetric/Cipher/Block/AESKeyWrap.cry:46]: Some of the test vectors in the test suite exercise failure. So, the only way to use those test vectors Is to return the tuple.

Comment (bboston7) [reply Primitive/Symmetric/Cipher/Block/AESKeyWrap.cry]: Thanks, I changed this to return a tuple instead.  I also added the KW-AD-AES256 tests from the Cryptol course repo ensure the tests also cover validation failure.

**Comment by bboston7** [comment]
I just added a padded keywrap spec.

File: Primitive/Symmetric/Cipher/Block/AESKeyWrapPadded.cry [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    iv' : [4][8]
+    size : [4][8]
+    P : [n-8][8]
+    iv' # size # P = S
+    // Validation check
+    valid =
+      // Most significant bytes of IV should match `iv`
+      (iv' == iv) &&
+      // n - 16 < size <= n - 8
+      ((((zero : [32 - (width n)]) # `n) - 16) < (join size)) &&
```
Comment (andreistefanescu) [inline Primitive/Symmetric/Cipher/Block/AESKeyWrapPadded.cry:84]: does ```((zero : [32 - (width n)]) # `n)``` -> ```(`n : [32])``` work?

**Review by andreistefanescu — APPROVED** [review]
Looks good!

File: Primitive/Symmetric/Cipher/Block/AESKeyWrapPadded.cry (hunk 2 above)
Comment (bboston7) [reply Primitive/Symmetric/Cipher/Block/AESKeyWrapPadded.cry]: Yep!  Don't know why I missed that as a possible simplification.  Thanks.

**Comment by bboston7** [comment]
I'm closing this PR because I manually merged the changes into `sha-imperative`.  The merge required a rebase to avoid clobbering Andre's AES changes and I didn't want to rebase `aes-keywrap` because others may already be using the branch.

**Pull request closed without merging**